rpcserver.go

```diff
@@ -6952,3 +6956,110 @@ func (r *rpcServer) FundingStateStep(ctx context.Context,
 	// current state?
 	return &lnrpc.FundingStateStepResp{}, nil
 }
+
+// RegisterRPCMiddleware adds a new gRPC middleware to the interceptor chain. A
+// gRPC middleware is software component external to lnd that aims to add
+// additional business logic to lnd by observing/intercepting/validating
+// incoming gRPC client requests and (if needed) replacing/overwriting outgoing
+// messages before they're sent to the client. When registering the middleware
+// must identify itself and indicate what custom macaroon caveats it wants to
+// be responsible for. Only requests that contain a macaroon with that specific
+// custom caveat are then sent to the middleware for inspection. As a security
+// measure, _no_ middleware can intercept requests made with _unencumbered_
+// macaroons!
+func (r *rpcServer) RegisterRPCMiddleware(
+	stream lnrpc.Lightning_RegisterRPCMiddlewareServer) error {
+
+	// This is a security critical functionality and needs to be enabled
+	// specifically by the user.
+	if !r.cfg.RPCMiddleware.Enable {
+		return fmt.Errorf("RPC middleware not enabled in config")
+	}
+
+	// When registering a middleware the first message being sent from the
+	// middleware must be a registration message containing its name and the
+	// custom caveat it wants to register for.
+	var (
+		registerMsg *lnrpc.MiddlewareRegistration
+		errChan     = make(chan error)
+	)
+	ctxc, cancel := context.WithTimeout(
+		stream.Context(), r.cfg.RPCMiddleware.InterceptTimeout,
+	)
+	defer cancel()
+
+	// Read the first message in a goroutine because the Recv method blocks
+	// until the message arrives.
+	go func() {
+		msg, err := stream.Recv()
+		if err != nil {
+			select {
+			case errChan <- err:
+			case <-ctxc.Done():
+			case <-r.quit:
+			}
+
+			return
+		}
+
+		registerMsg = msg.GetRegister()
+		close(errChan)
+	}()
+
+	// Wait for the initial message to arrive or time out if it takes too
+	// long.
+	select {
+	case err := <-errChan:
+		if err != nil {
+			return fmt.Errorf("error receiving initial middleware "+
+				"registration message: %v", err)
+		}
+
+		if registerMsg == nil {
+			return fmt.Errorf("invalid initial middleware " +
+				"registration message")
+		}
+
+	case <-ctxc.Done():
+		return ctxc.Err()
+
+	case <-r.quit:
+		return ErrServerShuttingDown
+	}
+
+	// Make sure the registration is valid.
+	const nameMinLength = 5
+	if len(registerMsg.MiddlewareName) < nameMinLength {
+		return fmt.Errorf("invalid middleware name, use descriptive "+
+			"name of at least %d characters", nameMinLength)
+	}
+
+	readOnly := registerMsg.ReadOnlyMode
```

Comment:
Do we need a nil check in case the middleware doesn't send a register msg first?

Reply:
Yes, that's done a few lines above.

Reply:
oops, meant to delete this!